grammar.py

```diff
@@ -204,165 +156,90 @@
-                charset |= category_to_charset(subpattern[1])
-            else:
-                raise NotImplementedError(f"I don't know how to generate examples of {subpattern[0]}")
-        result = bytes([random.choice(list(negate_charset(charset) if need_to_negate else charset))])
-    elif node_type == BRANCH:
-        result = helper(random.choice(node_value[1]))
-    elif node_type == ANY:
-        result = helper([(IN, [(LITERAL, b) for b in ALL_CHARSET])])
-    else:
-        raise NotImplementedError(f"I don't know how to generate examples of {node_type}")
-
-    return result + helper(parse_tree[1:])
-
-
-def generate_random_matching_input(pattern: bytes | str) -> bytes:
-    return helper(re_parse(pattern))
-
-
-def normalize_match(groupdict: dict[str, bytes | Any]) -> dict[str, bytes | Any]:
-    normalized_dict: dict[str, bytes | Any] = {}
-    for rule_name, rule_match in groupdict.items():
-        normalized_name: str = rule_name.removesuffix(REL_REF_EXT)
-        if normalized_name not in normalized_dict:
-            normalized_dict[normalized_name] = rule_match
-        elif normalized_dict[normalized_name] is None:
-            normalized_dict[normalized_name] = rule_match
-    return normalized_dict
-
-
-def get_rules_to_fill(groupdict: dict[str, bytes | Any]) -> list[str]:
-    rules_to_fill: list[str] = []
-    if any(
-        groupdict[rule_name] is not None
-        for rule_name in ("path_absolute", "path_empty", "path_rootless", "path_noscheme")
+def generate_random_match(rule_name: str) -> bytes:
+    return re_generate.generate_random_match_from_pattern(grammar_rules[rule_name])
+
+
+INSERTABLE_RULES: list[str] = ["query", "fragment", "scheme"]
+INSERTABLE_RULES_WITH_HOST: list[str] = ["port", "userinfo"]
+
+
+def grammar_insert(b: bytes) -> bytes:
+    match: re.Match[bytes] | None = re.match(URI_RE, b)
+    if match is None:
+        match = re.match(RELATIVE_REF_RE, b)
+    if match is None:
+        raise ValueError("Mutation precondition didn't hold.")
+
+    rules_to_fill: list[str] = [
+        s
+        for s in INSERTABLE_RULES + (INSERTABLE_RULES_WITH_HOST if match["host"] is not None else [])
+        if match[s] is None
+    ]
+
+    if len(rules_to_fill) == 0:
+        raise ValueError("Mutation precondition didn't hold.")
+
+    rule_name: str = random.choice(rules_to_fill)
+    return serialize(match.groupdict() | {rule_name: generate_random_match(rule_name)})
+
+
+def grammar_replace(b: bytes) -> bytes:
+    match: re.Match[bytes] | None = re.match(URI_RE, b)
+    if match is None:
+        match = re.match(RELATIVE_REF_RE, b)
+    if match is None:
+        raise ValueError("Mutation precondition didn't hold.")
+
+    groupdict: dict[str, bytes | None] = match.groupdict()
+    rule_name: str = random.choice([r for r in groupdict if groupdict[r] is not None])
+    return serialize(groupdict | {rule_name: generate_random_match(rule_name)})
+
+
+def grammar_delete(b: bytes) -> bytes:
```

Comment:
Have to ensure that we don't delete the last rule just like we can't delete the last byte. This can result in an empty input rn.

Reply:
Good catch; we'll definitely have to put in a check. I think the easiest thing to do is just raise ValueError if the result is b"", rather than reasoning about rules, which may or may not have delimiters when serialized, etc.